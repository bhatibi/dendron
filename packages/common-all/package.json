{
  "name": "@dendronhq/common-all",
  "version": "0.13.2-alpha.0",
  "description": "common-all",
  "license": "GPLv3",
  "repository": {
    "type": "git",
    "url": "ssh://git@github.com/dendronhq/dendron.git",
    "directory": "packages/common-all"
  },
  "author": {
    "name": "Kevin Lin",
    "email": "kevin@thence.io",
    "url": "https://kevinslin.com"
  },
  "keywords": [
    "common"
  ],
  "files": [
    "lib"
  ],
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "scripts": {
    "clean": "rimraf lib && rimraf coverage",
    "format": "prettier --ignore-path='../../.prettierignore' --write .",
    "lint": "echo stub",
    "prepublishOnly": "npm run build",
    "prebuild": "npm run clean && npm run format && npm run lint && echo Using TypeScript && tsc --version",
    "build": "npm run compile",
    "compile": "tsc -p tsconfig.build.json",
    "test": "npx jest",
    "coverage": "jest --coverage",
    "watch": "npm run compile -- --watch",
    "test:unit": "jest ",
    "test:unit:debug": "NODE_ENV=test node --inspect-brk node_modules/.bin/jest --runInBand",
    "test:watch": "jest --watch"
  },
  "dependencies": {
    "@types/jest": "^23.3.3",
    "@types/lodash": "^4.14.152",
    "@types/minimatch": "^3.0.3",
    "@types/moment": "^2.13.0",
    "@types/node": "13.11.0",
    "@types/pino": "^6.0.1",
    "@types/unist": "^2.0.3",
    "@types/uuid": "^8.0.0",
    "@types/yamljs": "^0.2.31",
    "dropbox": "^4.0.30",
    "gray-matter": "^4.0.2",
    "isomorphic-fetch": "^2.2.1",
    "js-yaml": "^3.14.0",
    "lodash": "^4.17.15",
    "minimatch": "^3.0.4",
    "moment": "^2.27.0",
    "pino": "^6.3.1",
    "request": "^2.88.2",
    "request-promise": "^4.2.6",
    "uuid": "^8.1.0",
    "vscode-uri": "^2.1.2",
    "yamljs": "^0.3.0"
  },
  "devDependencies": {
    "@dendronhq/common-test-utils": "^0.13.2-alpha.0",
    "@types/js-yaml": "^3.12.5",
    "@types/request-promise": "^4.1.46",
    "coveralls": "^3.0.2",
    "jest": "^23.6.0",
    "prettier": "2.0.5",
    "rimraf": "^2.6.2",
    "ts-jest": "^23.10.3",
    "ts-node": "^8.10.2",
    "tslint": "^5.11.0",
    "tslint-config-prettier": "^1.15.0",
    "typescript": "^3.1.1"
  },
  "engines": {
    "node": ">=12.0.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "jest": {
    "preset": "ts-jest"
  },
  "gitHead": "47c197f18329be43e02389fb4bd63267dfc07874"
}
