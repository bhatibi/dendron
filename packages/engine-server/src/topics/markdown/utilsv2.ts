import {
  DendronError,
  DLink,
  DNoteLoc,
  NotePropsV2,
  NoteUtilsV2,
} from "@dendronhq/common-all";
import _ from "lodash";
import remark from "remark";
import abbrPlugin from "remark-abbr";
import frontmatterPlugin from "remark-frontmatter";
import markdownParse from "remark-parse";
import {
  DendronLinksOpts,
  dendronLinksPlugin,
  WikiLinkNote,
} from "./plugins/dendronLinksPlugin";
import {
  root as mdastRoot,
  paragraph,
  text,
  heading,
  brk,
} from "mdast-builder";

const selectAll = require("unist-util-select").selectAll;

export class ParserUtilsV2 {
  static createWikiLinkRE(opts?: { oldLink: string }) {
    const { oldLink } = opts || {};
    if (oldLink) {
      const match = ParserUtilsV2.escapeForRegExp(oldLink);
      return `\\[\\[\\s*?(.*\\|)?\\s*${match}\\s*\\]\\]`;
    }
    return "\\[\\[\\s*?(.*\\|)?\\s*(?<name>.*)\\s*\\]\\]";
  }

  /**
   * - parse frontmatter
   * - parse wiki links
   * @param opts
   */
  static getRemark(opts?: { dendronLinksOpts: DendronLinksOpts }) {
    const { dendronLinksOpts } = _.defaults(opts, { dendronLinksOpts: {} });
    const errors: DendronError[] = [];
    return remark()
      .data("errors", errors)
      .use(markdownParse, { gfm: true })
      .use(abbrPlugin)
      .use(frontmatterPlugin, ["yaml"])
      .use(dendronLinksPlugin, dendronLinksOpts)
      .use({ settings: { listItemIndent: "1", fences: true } });
  }

  static findLinks({ note }: { note: NotePropsV2 }): DLink[] {
    const content = note.body;
    let remark = ParserUtilsV2.getRemark();
    //const out = remark.processSync(content);
    let out = remark.parse(content);
    let out2: WikiLinkNote[] = selectAll("wikiLink", out);
    const dlinks = out2.map(
      (m: WikiLinkNote) =>
        ({
          type: "wiki",
          from: NoteUtilsV2.toLoc(note),
          original: m.value,
          value: m.value,
          alias: m.data.alias,
          pos: { start: m.position?.start.offset, end: m.position?.end.offset },
          to: {
            fname: m.value,
          },
        } as DLink)
    );
    return dlinks as DLink[];
  }

  static escapeForRegExp(value: string) {
    return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  }

  static genMDError(opts: { msg: string; title: string }) {
    const { msg, title } = opts;
    return mdastRoot([
      heading(3, text(title)),
      paragraph([paragraph(text(msg)), brk]),
    ]);
  }

  static async replaceLinks(opts: {
    content: string;
    from: DNoteLoc;
    to: DNoteLoc;
  }) {
    const { content, from, to } = opts;
    let remark = ParserUtilsV2.getRemark({
      dendronLinksOpts: {
        replaceLink: { from, to },
      },
    });
    const out = await remark.process(content);
    return out.toString();
  }
}
