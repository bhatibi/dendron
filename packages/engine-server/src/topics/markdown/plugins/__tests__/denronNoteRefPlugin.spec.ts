import { DendronError, DEngineClientV2 } from "@dendronhq/common-all";
import { createLogger, EngineTestUtils } from "@dendronhq/common-server";
import { NodeTestUtilsV2 } from "@dendronhq/common-test-utils";
import _ from "lodash";
import { FileStorageV2 } from "../../../../drivers/file/storev2";
import { DendronEngineV2 } from "../../../../enginev2";
import { ParserUtilsV2 } from "../../utilsv2";
import {
  dendronNoteRefPlugin,
  DendronNoteRefPluginOpts,
} from "../dendronNoteRefPlugin";

function getProcessor(opts: DendronNoteRefPluginOpts) {
  return ParserUtilsV2.getRemark().use(dendronNoteRefPlugin, opts);
}
let LOGGER = createLogger("engine-server.test.log");

describe("basic", () => {
  describe("parse", () => {
    const opts = {
      renderWithOutline: true,
      replaceRefOpts: {
        forNoteRefInSite: true,
      },
      engine: {} as any,
    } as DendronNoteRefPluginOpts;

    test("init", () => {
      const resp = getProcessor(opts).parse(`((ref:[[foo.md]]))`);
      expect(resp).toMatchSnapshot();
      // child1 paragraph, child2 link
      // @ts-ignore
      expect(resp.children[0].children[0].data.link).toEqual({
        type: "file",
        name: "foo",
        anchorStart: undefined,
        anchorEnd: undefined,
      });
    });

    test("without suffix", () => {
      const resp = getProcessor(opts).parse(`((ref:[[foo]]))`);
      expect(resp).toMatchSnapshot();
      // child1 paragraph, child2 link
      // @ts-ignore
      expect(resp.children[0].children[0].data.link).toEqual({
        type: "file",
        name: "foo",
        anchorStart: undefined,
        anchorEnd: undefined,
      });
    });

    test("with start anchor", () => {
      const resp = getProcessor(opts).parse(`((ref:[[foo]]#h1))`);
      expect(resp).toMatchSnapshot();
      // child1 paragraph, child2 link
      // @ts-ignore
      expect(resp.children[0].children[0].data.link).toEqual({
        type: "file",
        name: "foo",
        anchorStart: "h1",
        anchorEnd: undefined,
      });
    });

    test("with start and end", () => {
      const resp = getProcessor(opts).parse(`((ref:[[foo]]#h1:#h2))`);
      expect(resp).toMatchSnapshot();
      // child1 paragraph, child2 link
      // @ts-ignore
      expect(resp.children[0].children[0].data.link).toEqual({
        type: "file",
        name: "foo",
        anchorStart: "h1",
        anchorEnd: "h2",
      });
    });

    test("doesn't parse inline code block", () => {
      const resp = getProcessor(opts).parse("`((ref:[[foo.md]]))`");
      expect(resp).toMatchSnapshot("bond");
      // @ts-ignore
      expect(resp.children[0].children[0].type).toEqual("inlineCode");
    });
    test.skip("doesn't parse code block", () => {});
  });

  describe("stingify", () => {
    let root: string;
    let opts: DendronNoteRefPluginOpts;
    let engine: DEngineClientV2;
    let vaultDir: string;

    beforeEach(async () => {
      root = await EngineTestUtils.setupVault({
        initDirCb: async (_vaultDir: string) => {
          vaultDir = _vaultDir;
          await NodeTestUtilsV2.createNote({
            vaultDir,
            noteProps: {
              fname: "daily.tasks",
              body: ["", `# Tasks`, "task1", "task2"].join("\n"),
            },
          });
        },
      });
      engine = new DendronEngineV2({
        vaults: [root],
        forceNew: true,
        store: new FileStorageV2({ vaults: [root], logger: LOGGER }),
        mode: "fuzzy",
        logger: LOGGER,
      });

      opts = {
        renderWithOutline: false,
        replaceRefOpts: {},
        engine,
      } as DendronNoteRefPluginOpts;
    });

    test("basic", async () => {
      await engine.init();
      const out = getProcessor(opts)
        .processSync(`((ref:[[daily.tasks]]))`)
        .toString();
      expect(out).toMatchSnapshot();
      expect(out.indexOf("task1") >= 0).toBeTruthy();
    });

    test("basic block", async () => {
      await engine.init();
      const out = getProcessor(opts)
        .processSync(
          `# Foo Bar
((ref:[[daily.tasks]]))`
        )
        .toString();
      expect(out).toMatchSnapshot();
      expect(out.indexOf("task1") >= 0).toBeTruthy();
    });

    test("basic block with fm", async () => {
      await NodeTestUtilsV2.createNote({
        vaultDir,
        noteProps: {
          fname: "daily.tasks",
          body: ["---", "id: foo", "---", `# Tasks`, "task1", "task2"].join(
            "\n"
          ),
        },
      });
      const out = getProcessor(opts)
        .processSync(
          `# Foo Bar
((ref:[[daily.tasks]]))`
        )
        .toString();
      expect(out).toMatchSnapshot();
      expect(out.indexOf("task1") >= 0).toBeTruthy();
      expect(out.indexOf("---") >= 0).toBeFalsy();
    });

    test("basic block with header and start ", async () => {
      const txt = [
        "---",
        "id: foo",
        "---",
        `# Tasks`,
        "## Header1",
        "task1",
        "## Header2",
        "task2",
      ];
      await NodeTestUtilsV2.createNote({
        vaultDir,
        noteProps: {
          fname: "daily.tasks",
          body: txt.join("\n"),
        },
      });
      const out = getProcessor(opts)
        .processSync(
          `# Foo Bar
((ref:[[daily.tasks]]#Header2))`
        )
        .toString();
      expect(out).toMatchSnapshot();
      expect(out.indexOf("task1") >= 0).toBeFalsy();
      expect(out.indexOf("task2") >= 0).toBeTruthy();
    });

    test("basic block with header and start, start invalid ", async () => {
      const txt = [
        "---",
        "id: foo",
        "---",
        `# Tasks`,
        "## Header1",
        "task1",
        "## Header2",
        "task2",
      ];
      await NodeTestUtilsV2.createNote({
        vaultDir,
        noteProps: {
          fname: "daily.tasks",
          body: txt.join("\n"),
        },
      });
      const out = getProcessor(opts)
        .processSync(
          `# Foo Bar
((ref:[[daily.tasks]]#badheader))`
        )
        .toString();
      expect(out).toMatchSnapshot();
      expect(out.indexOf("Note Ref Error") >= 0).toBeTruthy();
      expect(out.indexOf("badheader not found") >= 0).toBeTruthy();
    });

    test("basic block with header, start and end, end invalid ", async () => {
      const txt = [
        "---",
        "id: foo",
        "---",
        `# Tasks`,
        "## Header1",
        "task1",
        "## Header2",
        "task2",
        "<div class='bar'>",
        "BOND",
        "</div>",
      ];
      await NodeTestUtilsV2.createNote({
        vaultDir,
        noteProps: {
          fname: "daily.tasks",
          body: txt.join("\n"),
        },
      });
      const out = getProcessor(opts)
        .processSync(
          `# Foo Bar
((ref:[[daily.tasks]]#Header1:#badheader))`
        )
        .toString();
      expect(out).toMatchSnapshot();
      expect(out.indexOf("Note Ref Error") >= 0).toBeTruthy();
      expect(out.indexOf("badheader not found") >= 0).toBeTruthy();
    });

    test("basic block with header and start, offset ", async () => {
      const txt = [
        "---",
        "id: foo",
        "---",
        `# Tasks`,
        "## Header1",
        "task1",
        "## Header2",
        "task2",
      ];
      await NodeTestUtilsV2.createNote({
        vaultDir,
        noteProps: {
          fname: "daily.tasks",
          body: txt.join("\n"),
        },
      });
      const out = getProcessor(opts)
        .processSync(
          `# Foo Bar
((ref:[[daily.tasks]]#Header2,1))`
        )
        .toString();
      expect(out).toMatchSnapshot();
      expect(out.indexOf("Header2") >= 0).toBeFalsy();
      expect(out.indexOf("task1") >= 0).toBeFalsy();
      expect(out.indexOf("task2") >= 0).toBeTruthy();
    });

    test("basic block with header, start and end ", async () => {
      const txt = [
        "---",
        "id: foo",
        "---",
        `# Tasks`,
        "## Header1",
        "task1",
        "## Header2",
        "task2",
        "<div class='bar'>",
        "BOND",
        "</div>",
      ];
      await NodeTestUtilsV2.createNote({
        vaultDir,
        noteProps: {
          fname: "daily.tasks",
          body: txt.join("\n"),
        },
      });
      const out = getProcessor(opts)
        .processSync(
          `# Foo Bar
((ref:[[daily.tasks]]#Header1:#Header2))`
        )
        .toString();
      expect(out).toMatchSnapshot();
      expect(out.indexOf("Header1") >= 0).toBeTruthy();
      expect(out.indexOf("task1") >= 0).toBeTruthy();
      expect(out.indexOf("task2") >= 0).toBeFalsy();
    });

    test("basic block with header, start and end, offset ", async () => {
      const txt = [
        "---",
        "id: foo",
        "---",
        `# Tasks`,
        "## Header1",
        "task1",
        "## Header2",
        "task2",
        "<div class='bar'>",
        "BOND",
        "</div>",
      ];
      await NodeTestUtilsV2.createNote({
        vaultDir,
        noteProps: {
          fname: "daily.tasks",
          body: txt.join("\n"),
        },
      });
      const out = getProcessor(opts)
        .processSync(
          `# Foo Bar
((ref:[[daily.tasks]]#Header1,1:#Header2))`
        )
        .toString();
      expect(out).toMatchSnapshot();
      expect(out.indexOf("Header1") >= 0).toBeFalsy();
      expect(out.indexOf("task1") >= 0).toBeTruthy();
      expect(out.indexOf("task2") >= 0).toBeFalsy();
    });

    test("basic block with wildcard as 1st elem ", async () => {
      const txt = [
        "---",
        "id: foo",
        "---",
        `# Tasks`,
        "## Header1",
        "task1",
        "## Header2",
        "task2",
        "<div class='bar'>",
        "BOND",
        "</div>",
      ];
      await NodeTestUtilsV2.createNote({
        vaultDir,
        noteProps: {
          fname: "daily.tasks",
          body: txt.join("\n"),
        },
      });
      const out = getProcessor(opts)
        .processSync(
          `# Foo Bar
((ref:[[daily.tasks]]#*,1:#header2))`
        )
        .toString();
      expect(out).toMatchSnapshot();
      expect(out.indexOf("Header1") >= 0).toBeTruthy();
      expect(out.indexOf("task1") >= 0).toBeTruthy();
      expect(out.indexOf("task2") >= 0).toBeFalsy();
    });

    test("basic block with wildcard as 2nd elem ", async () => {
      const txt = [
        "---",
        "id: foo",
        "---",
        `# Tasks`,
        "## Header1",
        "task1",
        "## Header2",
        "task2",
        "<div class='bar'>",
        "BOND",
        "</div>",
      ];
      await NodeTestUtilsV2.createNote({
        vaultDir,
        noteProps: {
          fname: "daily.tasks",
          body: txt.join("\n"),
        },
      });
      const out = getProcessor(opts)
        .processSync(
          `# Foo Bar
((ref:[[daily.tasks]]#Header1,1:#*))`
        )
        .toString();
      expect(out).toMatchSnapshot();
      expect(out.indexOf("Header1") >= 0).toBeFalsy();
      expect(out.indexOf("task1") >= 0).toBeTruthy();
      expect(out.indexOf("task2") >= 0).toBeFalsy();
    });

    test("2 lvl recursion", async () => {
      await NodeTestUtilsV2.createNote({
        vaultDir,
        noteProps: {
          fname: "foo.one",
          body: ["# Foo.One", `((ref: [[foo.two]]))`].join("\n"),
        },
      });
      await NodeTestUtilsV2.createNote({
        vaultDir,
        noteProps: {
          fname: "foo.two",
          body: ["# Foo.Two", `blah`].join("\n"),
        },
      });
      const out = getProcessor(opts)
        .processSync(["# Foo", "((ref: [[foo.one]]))"].join("\n"))
        .toString();
      expect(out).toMatchSnapshot();
      _.every(["# Foo", "# Foo.One", "# Foo.Two"], (ent) => {
        expect(out.indexOf(ent) >= 0).toBeTruthy();
      });
    });

    test("3 levels of recursion", async () => {
      await NodeTestUtilsV2.createNote({
        vaultDir,
        noteProps: {
          fname: "foo.one",
          body: ["# Foo.One", `((ref: [[foo.two]]))`].join("\n"),
        },
      });
      await NodeTestUtilsV2.createNote({
        vaultDir,
        noteProps: {
          fname: "foo.two",
          body: ["# Foo.Two", `((ref: [[foo.three]]))`].join("\n"),
        },
      });
      await NodeTestUtilsV2.createNote({
        vaultDir,
        noteProps: {
          fname: "foo.three",
          body: ["# Foo.Three", `Three`].join("\n"),
        },
      });
      const out = getProcessor(opts)
        .processSync(["# Foo", "((ref: [[foo.one]]))"].join("\n"))
        .toString();
      expect(out).toMatchSnapshot();
      _.every(
        [
          "# Foo",
          "# Foo.One",
          "# Foo.Two",
          "ERROR: Too many nested note references",
        ],
        (ent) => {
          expect(out.indexOf(ent) >= 0).toBeTruthy();
        }
      );
    });

    test("ref with ref, inf recursion", async () => {
      await NodeTestUtilsV2.createNote({
        vaultDir,
        noteProps: {
          fname: "foo.one",
          body: ["# Foo.One", `((ref: [[foo.two]]))`].join("\n"),
        },
      });
      await NodeTestUtilsV2.createNote({
        vaultDir,
        noteProps: {
          fname: "foo.two",
          body: ["# Foo.Two", `((ref: [[foo.one]]))`].join("\n"),
        },
      });
      const out = getProcessor(opts)
        .processSync(["# Foo", "((ref: [[foo.one]]))"].join("\n"))
        .toString();
      expect(out).toMatchSnapshot();
      _.every(
        [
          "# Foo",
          "# Foo.One",
          "# Foo.Two",
          "ERROR: Too many nested note references",
        ],
        (ent) => {
          expect(out.indexOf(ent) >= 0).toBeTruthy();
        }
      );
    });

    test("renderWithOutline", async () => {
      const txt = [
        "---",
        "id: foo",
        "---",
        `# Tasks`,
        "## Header1",
        "task1",
        "## Header2",
        "task2",
        "<div class='bar'>",
        "BOND",
        "</div>",
      ];
      await NodeTestUtilsV2.createNote({
        vaultDir,
        noteProps: {
          fname: "daily.tasks",
          body: txt.join("\n"),
        },
      });
      const out = getProcessor({ ...opts, renderWithOutline: true })
        .processSync(
          `# Foo Bar
((ref:[[daily.tasks]]#Header1:#Header2))`
        )
        .toString();
      expect(out).toMatchSnapshot();
      expect(out.indexOf("Header1") >= 0).toBeTruthy();
      expect(out.indexOf("task1") >= 0).toBeTruthy();
      expect(out.indexOf("task2") >= 0).toBeFalsy();
    });

    test("missing ref", async () => {
      const proc = getProcessor({ ...opts, renderWithOutline: true });
      const out = proc
        .processSync(
          `# Foo Bar
((ref:[[missing]]#Header1:#Header2))`
        )
        .toString();
      expect(out).toMatchSnapshot();
      const errors = proc.data("errors") as DendronError[];
      expect(errors).toMatchSnapshot();
      expect(errors[0].msg).toEqual("missing not found");
      expect(out.indexOf("missing not found") >= 0).toBeTruthy();
    });
  });
});
