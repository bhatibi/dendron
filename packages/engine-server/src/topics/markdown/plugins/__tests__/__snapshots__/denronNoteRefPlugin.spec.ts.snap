// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic parse doesn't parse inline code block: bond 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 21,
              "line": 1,
              "offset": 20,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "inlineCode",
          "value": "((ref:[[foo.md]]))",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 21,
          "line": 1,
          "offset": 20,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 21,
      "line": 1,
      "offset": 20,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`basic parse init 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "data": Object {
            "alias": "ref:[[foo.md]]",
            "exists": false,
            "hChildren": Array [
              Object {
                "type": "text",
                "value": "ref:[[foo.md]]",
              },
            ],
            "hName": "a",
            "hProperties": Object {
              "className": "internal new",
              "href": "ref:[[foo",
            },
            "link": Object {
              "anchorEnd": undefined,
              "anchorStart": undefined,
              "name": "foo",
              "type": "file",
            },
            "permalink": "ref:[[foo",
          },
          "position": Position {
            "end": Object {
              "column": 19,
              "line": 1,
              "offset": 18,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "refLink",
          "value": "ref:[[foo.md]]",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 19,
          "line": 1,
          "offset": 18,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 19,
      "line": 1,
      "offset": 18,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`basic parse with start anchor 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "data": Object {
            "alias": "ref:[[foo]]#h1",
            "exists": false,
            "hChildren": Array [
              Object {
                "type": "text",
                "value": "ref:[[foo]]#h1",
              },
            ],
            "hName": "a",
            "hProperties": Object {
              "className": "internal new",
              "href": "ref:[[foo]]#h1",
            },
            "link": Object {
              "anchorEnd": undefined,
              "anchorStart": "h1",
              "name": "foo",
              "type": "file",
            },
            "permalink": "ref:[[foo]]#h1",
          },
          "position": Position {
            "end": Object {
              "column": 19,
              "line": 1,
              "offset": 18,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "refLink",
          "value": "ref:[[foo]]#h1",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 19,
          "line": 1,
          "offset": 18,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 19,
      "line": 1,
      "offset": 18,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`basic parse with start and end 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "data": Object {
            "alias": "ref:[[foo]]#h1:#h2",
            "exists": false,
            "hChildren": Array [
              Object {
                "type": "text",
                "value": "ref:[[foo]]#h1:#h2",
              },
            ],
            "hName": "a",
            "hProperties": Object {
              "className": "internal new",
              "href": "ref:[[foo]]#h1:#h2",
            },
            "link": Object {
              "anchorEnd": "h2",
              "anchorStart": "h1",
              "name": "foo",
              "type": "file",
            },
            "permalink": "ref:[[foo]]#h1:#h2",
          },
          "position": Position {
            "end": Object {
              "column": 23,
              "line": 1,
              "offset": 22,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "refLink",
          "value": "ref:[[foo]]#h1:#h2",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 23,
          "line": 1,
          "offset": 22,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 23,
      "line": 1,
      "offset": 22,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`basic parse without suffix 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "data": Object {
            "alias": "ref:[[foo]]",
            "exists": false,
            "hChildren": Array [
              Object {
                "type": "text",
                "value": "ref:[[foo]]",
              },
            ],
            "hName": "a",
            "hProperties": Object {
              "className": "internal new",
              "href": "ref:[[foo]]",
            },
            "link": Object {
              "anchorEnd": undefined,
              "anchorStart": undefined,
              "name": "foo",
              "type": "file",
            },
            "permalink": "ref:[[foo]]",
          },
          "position": Position {
            "end": Object {
              "column": 16,
              "line": 1,
              "offset": 15,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "refLink",
          "value": "ref:[[foo]]",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 16,
          "line": 1,
          "offset": 15,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 16,
      "line": 1,
      "offset": 15,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`basic stingify 2 lvl recursion 1`] = `
"# Foo

# Foo.One

# Foo.Two

blah


"
`;

exports[`basic stingify 3 levels of recursion 1`] = `
"# Foo

# Foo.One

# Foo.Two

ERROR: Too many nested note references


"
`;

exports[`basic stingify basic 1`] = `
"# Tasks

task1
task2


"
`;

exports[`basic stingify basic block 1`] = `
"# Foo Bar

# Tasks

task1
task2


"
`;

exports[`basic stingify basic block with fm 1`] = `
"# Foo Bar

# Tasks

task1
task2


"
`;

exports[`basic stingify basic block with header and start  1`] = `
"# Foo Bar

## Header2

task2


"
`;

exports[`basic stingify basic block with header and start, offset  1`] = `
"# Foo Bar


task2


"
`;

exports[`basic stingify basic block with header and start, start invalid  1`] = `
"# Foo Bar

### Note Ref Error

start anchor badheader not found  


"
`;

exports[`basic stingify basic block with header, start and end  1`] = `
"# Foo Bar

## Header1

task1


"
`;

exports[`basic stingify basic block with header, start and end, end invalid  1`] = `
"# Foo Bar

### Note Ref Error

end anchor badheader not found  


"
`;

exports[`basic stingify basic block with header, start and end, offset  1`] = `
"# Foo Bar


task1


"
`;

exports[`basic stingify basic block with wildcard as 1st elem  1`] = `
"# Foo Bar


## Header1

task1


"
`;

exports[`basic stingify basic block with wildcard as 2nd elem  1`] = `
"# Foo Bar


task1


"
`;

exports[`basic stingify missing ref 1`] = `
"# Foo Bar

### Note Ref Error

missing not found  


"
`;

exports[`basic stingify missing ref 2`] = `
Array [
  [Error: missing not found],
]
`;

exports[`basic stingify ref with ref, inf recursion 1`] = `
"# Foo

# Foo.One

# Foo.Two

ERROR: Too many nested note references


"
`;

exports[`basic stingify renderWithOutline 1`] = `
"# Foo Bar



<div class=\\"portal-container\\">
<div class=\\"portal-head\\">
<div class=\\"portal-backlink\\" >
<div class=\\"portal-title\\">From <span class=\\"portal-text-title\\">daily.tasks</span></div>
<a href=\\"daily.tasks.html\\" class=\\"portal-arrow\\">Go to text <span class=\\"right-arrow\\">→</span></a>
</div>
</div>
<div id=\\"portal-parent-anchor\\" class=\\"portal-parent\\" markdown=\\"1\\">
<div class=\\"portal-parent-fader-top\\"></div>
<div class=\\"portal-parent-fader-bottom\\"></div>        
  
## Header1

task1



</div>    
</div>

"
`;
